### PR TITLE
Add smart contract wallet

### PR DESCRIPTION
**Pull request by DOBEN**
## Purpose

closes https://github.com/Concordium/concordium-rust-smart-contracts/issues/412

Next: Update the CIS5 standard proposal document to match the reference implementation:
https://github.com/Concordium/concordium-update-proposals/pull/60
e.g. the service fee is deposited to the `public_key` of the service fee recipient in this implementation to avoid too much dust amounts being sent to addresses.

## Changes

Add smart contract wallet reference implementation.

### CODE REVIEW

**Review by Bargsteen — APPROVED** [review]
Nice work ;)

File: examples/cis5-smart-contract-wallet/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,1465 @@
...
+        }
+    }
+}
+
+/// The contract state.
+#[derive(Serial, DeserialWithState)]
+#[concordium(state_parameter = "S")]
+struct State<S = StateApi> {
+    /// The token balances stored in the state.
+    token_balances:  StateMap<ContractAddress, ContractAddressState<S>, S>,
```
[the remark this answers is not in the document]
Comment (DOBEN) [reply examples/cis5-smart-contract-wallet/src/lib.rs:216]: I got a warning by `clippy` that the type is too complex.

File: examples/cis5-smart-contract-wallet/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,1465 @@
...
+//! holder of the corresponding private key, ensuring self-custodial control
+//! over the assets.
+//!
+//! Transfers of native currency and CIS-2 token balances do not require
+//! on-chain transaction submissions. Instead, the holder of the corresponding
+//! private key can generate a valid signature and identify a third party to
+//! submit the transaction on-chain, potentially incentivizing the third-party
+//! involvement through a service fees. The `withdraw` and `internalTransfer`
+//! functions transfer (authorized as part of the message that was signed) the
+//! amount of service fee to the service fee recipient's public key.
```
[the remark this answers is not in the document]
Comment (DOBEN) [reply examples/cis5-smart-contract-wallet/src/lib.rs:20]: The service fee recipient is a `public key` in this smart contract wallet. The design decision was to not have too many small dust transfers. After a service fee recipient has collected enough fees, they can transfer their assets out of this smart contract wallet.

Rephrased the sentence.

File: examples/cis5-smart-contract-wallet/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,1465 @@
...
+        cis2_token_contract_address: &ContractAddress,
+        public_key: &PublicKeyEd25519,
+    ) -> ContractTokenAmount {
+        self.token_balances
+            .get(cis2_token_contract_address)
+            .map(|a| {
+                a.balances
+                    .get(token_id)
+                    .map(|b| {
+                        b.get(public_key).map(|c| *c).unwrap_or_else(|| TokenAmountU256(0.into()))
```
[the remark this answers is not in the document]
Comment (DOBEN) [reply examples/cis5-smart-contract-wallet/src/lib.rs:264]: We don't have good `from` traits for the type `ContractTokenAmount`, so I kept the current way of initializing the `TokenAmountU256` .

File: examples/cis5-smart-contract-wallet/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,1465 @@
...
+    /// The signature.
+    pub signature: SignatureEd25519,
+    /// The message being signed.
+    pub message:   WithdrawMessage,
+}
+
+/// The parameter type for the contract functions
+/// `withdrawNativeCurrency/withdrawCis2Tokens`.
+#[derive(Serialize, SchemaType)]
+#[concordium(transparent)]
```
[the remark this answers is not in the document]
Comment (DOBEN) [reply examples/cis5-smart-contract-wallet/src/lib.rs:619]: Could you explain why this is needed please?

**Comment by DOBEN** [comment]
I decided to change the `from_address` in the receiveHook function call in the cis2_multi contract. This chain seems to make more sense in terms that the token is moved form one address/public_key to another. The `DepositCIs2Tokens` is logged before the `MintEvent` but both events are in the same atomic transaction.

MintEvent -> to: `to_address`
DepositCIs2Tokens -> from `to_address` to `alice_public_key`.
https://github.com/Concordium/concordium-rust-smart-contracts/blob/e5f590fc38c78fa560055ddaec1884262a77cc31/examples/cis2-multi/src/lib.rs#L1044

File: examples/cis5-smart-contract-wallet/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (limemloh) [reply examples/cis5-smart-contract-wallet/src/lib.rs:585]: We discussed this offline

**Review by limemloh — APPROVED** [review]
LGTM :+1:

**Pull request merged**